Filling in threat mitigations in privacy-security-explainer.md (#746)
immersive-web/webxr: privacy-security-explainer.md

# Protected functionality
The sensitive information exposed via WebXR can be divided into categories that share threat profiles and necessary protections against those threats.

## Immersiveness
Users must be in control of when immersive sessions are created because the creation causes invasive changes on a user's machine. For example, starting an immersive session will engage the XR device sensors, take over access to the device's display, and begin presentation which may terminate another application's access to the XR hardware. It may also incur significant power or performance overhead on some systems or trigger the launching of a status tray or storefront.

Developers indicate the desire to create an immersive session by passing `immersive-vr` or `immersive-ar` into `xr.requestSession()`. 

```js
// VR button click handler
function onVRClick() {
  xr.requestSession('immersive-vr').then(onVRSessionCreated);
}

//AR button click handler
function onARClick() {
  xr.requestSession('immersive-ar').then(onARSessionCreated);
}
```

In response, the UA must ensure that:

> * The function was invoked in response to a [user activation](#user-activation)

Reviewer: > in response to a user activation

Nit: I *think* that this is probably correct grammer, but it sounds weird to me. :) Would "user activation event" work better?

* The request originates from a [trustworthy document and origin](#trustworthy-documents-and-origins)
* The request originates from a document that is [visible and has focus](#visibility-and-focus)
* The request originates from a document allowed to use the WebXR [feature policy](#feature-policy) as well as the [underlying sensors' feature policies](#underlying-sensors-feature-policies)
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

If these requirements are not met, the promise returned from `requestSession()` must reject.

## Poses
### XRPose
When based on sensor data, calls to `XRFrame.getPose()` will expose sensitive information that may be misused in a number of ways, including input sniffing, gaze tracking, or fingerprinting.

Developers indicate the desire for `XRPose` data by calling `XRFrame.getPose()`.

```js
function onSessionRafCallback(XRFrame frame) {
    let motionControllerPose = frame.getPose(xrSession.inputSources[0], xrReferenceSpace);
}
```

Before returning poses based on sensor data, the UA must ensure that:

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent); alternatively, in cases where the user experience is not negatively affected, quantization may be applied* The request originates from the document which owns the `XRFrame`'s `XRSession`

Reviewer 2: > quantization may be applied* The request

Missing newline for the bullet point?

* The document is [visible and has focus](#visibility-and-focus)
* The `XRSession.visibility` is set to `visible` 
* **TODO** address issues [#696](https://github.com/immersive-web/webxr/issues/696) and [#724](https://github.com/immersive-web/webxr/issues/724)

### XRViewerPose

> The primary difference between `XRViewerPose` and `XRPose` is the inclusion of `XRView` information. More than one view may be present for a number of reasons. One example is a headset which will generally have two views, but may have more to accommodate greater than 180 degree field of views. Another example is a CAVE system. In all cases, when more than one view is present and the physical relationship between these views is configurable by the user, the relationship between these views is considered sensitive information as it can be used to fingerprint or profile the user.

Reviewer: > One example is a headset which will generally have two views

Nit: Missing a comma after "headset"?

Developers indicate the desire for `XRViewerPose` data by calling `XRFrame.getViewerPose()`.

```js
function onSessionRafCallback(XRFrame frame) {
    let viewerPose = frame.getViewerPose(xrReferenceSpace);
}
```

> As a result, when more than one view is present, the UA must additionally ensure that:

Reviewer: Feels like this should more clearly specify that all the restrictions that apply to XRPose still apply here. It hints at that by saying "additionally", but I feel it's far enough removed from the XRPose context that it's worth making explicit.

Reviewer 3: I agree; it's an inherited class but worth calling out that these requirements don't override the previous ones. I'd also suggest making this an ongoing (not one-time) requirement. Suggest:

```suggestion
For every call to getViewerPose(), before returning poses based on sensor data the UA must ensure that all of the requirements of getPose() are met. Further, when more than one view is present, the UA must ensure that:
```

* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

> * If the device supports configurable or factory-calibrated interpupillary distance that may vary from device to device, then the XRView transform data must be rounded to prevent fingerprinting. Specific precision for rounding is at the discretion of the user agent.

Reviewer 2: > factory-calibrated interpupillary distance

Consider rewording? The factory calibration could affect other data such as the view frustum or projection matrices, and in this case the calibrated value would be lens center distance, not interpupillary distance. Also, this is only an issue if these parameters affect the information exposed to WebXR.

Suggestion:

`If view data (view origins, frustum shape, or projection matrices) is affected by configurable or factory-calibrated parameters such as interpupillary distance or variations in screen geometry that may vary from device to device`

Reviewer: To piggyback on Klaus' point: if the IPD is statically set by the factory then that's presents a fingerprinting risk, and if it's user calibrated then that presents both a fingerprinting and a profiling risk. I wonder if that's something that can be cleanly expressed here?

}
```

In response, the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of a bounded reference space
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)
* The device is capable of bounded tracking

If these requirements are not met and `bounded-floor` is listed in `XRSessionInit.requiredFeatures` then the promise returned from `requestSession()` must be rejected. Otherwise, the promise may be fulfilled but future calls to `XRSession.requestReferenceSpace()` must fail when passed `bounded-floor`.

Once a session is created, developers may attempt to create a bounded reference space by passing `bounded-floor` into `XRSession.requestReferenceSpace()`.

```js
let xrReferenceSpace;
function onSessionCreated(session) {
  session.requestReferenceSpace('bounded-floor')
  .then((referenceSpace) => { xrReferenceSpace = referenceSpace; })
  .catch( (e) => { /* handle gracefully */ } );
}
```

In response, the UA must ensure that: 
* Bounded reference spaces are allowed to be created based on the restrictions above

> * Any group of `local`, `local-floor`, and `bounded-floor` reference spaces that are capable of being related to one another must share a common native origin; this restriction does not apply when `unbounded` reference spaces are also able to be created

Author: I'm actually not 100% certain this addresses the motion controller issue (and realized as much over the weekend) as there is no firm technical requirement that a motion controller share a native origin with anything.  Either way it might be useful to have a dedicated section for other `XRSpace` types and start it off with `XRInputSource`.  I'm gonna poke at this a bit this morning and see if I can come up with something good.

> * Any group of `local`, `local-floor`, and `bounded-floor` reference spaces that are capable of being related to one another must share a common native origin; this restriction does not apply when `unbounded` reference spaces are also able to be created

Author: Tracked in #752

* `XRBoundedReferenceSpace.boundsGeometry` must be [limited](#limiting) to a reasonable distance from the reference space's native origin; the suggested default distance is 15 meters in each direction
* Each point in the `XRBoundedReferenceSpace.boundsGeometry` must be [rounded](#rounding) sufficiently to prevent fingerprinting; rounding to the nearest 5cm is suggested
* The region represented by the rounded bounds geometry must be a subset of the original bounds

> * If the floor level is based on sensor data or is set to a non-default emulated value, the `y` value of the native origin must be [rounded](#rounding) sufficiently to prevent fingerprinting; rounding to the nearest 1cm is suggested

Author: I ended up making a slightly different change here.  Can you confirm that was sufficient or let me know if the proposed change is still necessary?

> * If the floor level is based on sensor data or is set to a non-default emulated value, the `y` value of the native origin must be [rounded](#rounding) sufficiently to prevent fingerprinting; rounding to the nearest 1cm is suggested

Reviewer 3: This one can be resolved, I'll suggest changes to the text in the new commit

* All `XRPose` and `XRViewerPose` 6DOF pose data computed using a `bounded-floor` reference space must be [limited](#limiting) to a reasonable distance beyond the `boundsGeometry` in all directions; the suggested distance is 1 meter beyond the bounds in all directions

If these requirements are not met, the promise returned from `XRSession.requestReferenceSpace()` must be rejected.

### Local and local-floor reference spaces
On devices which support 6-DOF tracking, `local` and `local-floor` reference spaces may be used to perform gait analysis, allowing user profiling and fingerprinting. In addition, because the `local-floor` reference spaces provide an established floor level, it may be possible for a site to infer the user’s height, allowing user profiling and fingerprinting.  

> When creating an `inline` session, developers indicate the desire for local viewer tracking at the time of session creation by adding `local` and/or `local-floor` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`. Creating an `immersive-vr` or `immersive-ar` session adds `local` and `local-floor` to `XRSessionInit.requiredFeatures` implicitly.

Author: I ended up separating the local and local floor sections to address this more effectively based on our conversation Friday.  So hopefully that will clarify things significantly. Should have updates posted shortly after lunch

```js
function onVRClick() {
  xr.requestSession('immersive-vr', { requiredFeatures: ['local-floor'] } )
  .then(onVRSessionCreated);
}

xr.requestSession('inline', { optionalFeatures: ['local'] } )
  .then(onInlineSessionCreated);
}
```

In response, the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of a local or local-floor reference space
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

> * The device is capable of local tracking

Author: I had been thinking it was sufficiently covered the in the section just below which describes the guidance for calls to `requestReferenceSpace()`.  Is there something more that needs addressing up here?

> * The device is capable of local tracking

Reviewer 3: New commit looks good on this point, this comment can be resolved

Developers indicate the desire for unbounded viewer tracking at the time of session creation by adding `unbounded` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`. 

```js
function onARClick() {
  xr.requestSession('immersive-ar', { requiredFeatures: ['unbounded'] } )
  .then(onARSessionCreated);
}
```

In response the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of an unbounded reference space
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implied consent](#implied-consent)
* The XR device is capable of unbounded tracking

If these requirements are not met and `unbounded` is listed in `XRSessionInit.requiredFeatures` then the promise returned from `requestSession()` must be rejected. Otherwise, the promise may be fulfilled but future calls to `XRSession.requestReferenceSpace()` must fail when passed `unbounded`.

> Once a session is created, developers may attempt to create a bounded reference space by passing `unbounded` into `XRSession.requestReferenceSpace()`.

Reviewer 4: Is this supposed to be 'create an unbounded reference space' or am I missing something?

Developers indicate the desire to create an immersive session by passing `immersive-vr` or `immersive-ar` into `xr.requestSession()`. 

```js
// VR button click handler
function onVRClick() {
  xr.requestSession('immersive-vr').then(onVRSessionCreated);
}

//AR button click handler
function onARClick() {
  xr.requestSession('immersive-ar').then(onARSessionCreated);
}
```

In response, the UA must ensure that:
* The function was invoked in response to a [user activation](#user-activation) event
* The request originates from a [trustworthy document and origin](#trustworthy-documents-and-origins)
* The request originates from a document that is [visible and has focus](#visibility-and-focus)
* The request originates from a document allowed to use the WebXR [feature policy](#feature-policy) as well as the [underlying sensors' feature policies](#underlying-sensors-feature-policies)

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo
```suggestion
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

* The `XRSession.visibility` is set to `visible` 
* **TODO** address issues [#696](https://github.com/immersive-web/webxr/issues/696) and [#724](https://github.com/immersive-web/webxr/issues/724)

### XRViewerPose
The primary difference between `XRViewerPose` and `XRPose` is the inclusion of `XRView` information. More than one view may be present for a number of reasons. One example is a headset, which will generally have two views, but may have more to accommodate greater than 180 degree field of views. Another example is a CAVE system. In all cases, when more than one view is present and the physical relationship between these views is configurable by the user, the relationship between these views is considered sensitive information as it can be used to fingerprint or profile the user.

Developers indicate the desire for `XRViewerPose` data by calling `XRFrame.getViewerPose()`.

```js
function onSessionRafCallback(XRFrame frame) {
    let viewerPose = frame.getViewerPose(xrReferenceSpace);
}
```

In addition to meeting the [`XRPose`](#xrpose) requirements, every call to `XRFrame.getViewerPose()` which will return more than one `XRView` must additionally ensure that:

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo in anchor:
```suggestion
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

* If `XRView` data is affected by a user-configured interpupillary distance, then the XRView data must be anonymized to prevent profiling as well as fingerprinting. Specific approaches to this are at the discretion of the user agent.

## Reference spaces
### Unbounded reference spaces
Unbounded reference spaces reveal the largest amount of spatial data and may result in user profiling and fingerprinting. For example, this data may enable determining user’s specific geographic location or to perform gait analysis.

Developers indicate the desire for unbounded viewer tracking at the time of session creation by adding `unbounded` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`. 

```js
function onARClick() {
  xr.requestSession('immersive-ar', { requiredFeatures: ['unbounded'] } )
  .then(onARSessionCreated);
}
```

In response the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of an unbounded reference space

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo in link
```suggestion
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

  .catch( (e) => { /* handle gracefully */ } );
}
```

### Bounded reference spaces
Bounded reference spaces, when sufficiently constrained in size, do not enable developers to determine geographic location. However, because the floor level is established and users are able to walk around, it may be possible for a site to infer the user’s height or perform gait analysis, allowing user profiling and fingerprinting. In addition, it may be possible perform fingerprinting using the bounds reported by a bounded reference space.

Developers indicate the desire for bounded viewer tracking at the time of session creation by adding `bounded-floor` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`. 

```js
function onVRClick() {
  xr.requestSession('immersive-vr', { requiredFeatures: ['bounded-floor'] } )
  .then(onVRSessionCreated);
}

xr.requestSession('inline', { optionalFeatures: ['bounded-floor'] } )
  .then(onInlineSessionCreated);
}
```

In response, the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of a bounded reference space

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo in link
```suggestion
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

If these requirements are not met, the promise returned from `XRSession.requestReferenceSpace()` must be rejected.

### Local-floor spaces
On devices which support 6-DOF tracking, `local-floor` reference spaces may be used to perform gait analysis, allowing user profiling and fingerprinting. In addition, because the `local-floor` reference spaces provide an established floor level, it may be possible for a site to infer the user’s height, allowing user profiling and fingerprinting.  

Developers indicate the desire for `local-floor` viewer tracking at the time of session creation by adding `local-floor` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`.

```js
function onVRClick() {
  xr.requestSession('immersive-vr', { requiredFeatures: ['local-floor'] } )
  .then(onVRSessionCreated);
}
```

In response, the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of a `local-floor` reference space

> * User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo in link
```suggestion
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

If these requirements are not met, the promise returned from `XRSession.requestReferenceSpace()` must be rejected.

### Local reference spaces
On devices which support 6-DOF tracking, `local` reference spaces may be used to perform gait analysis, allowing user profiling and fingerprinting.

When creating an `immersive-vr` or `immersive-ar` session, developers do not need to explicitly request the desire for `local` viewer tracking. However, this desire must be indicated when creating an `inline` session by adding `local` to either `XRSessionInit.requiredFeatures` or `XRSessionInit.optionalFeatures`. 

```js
xr.requestSession('inline', { optionalFeatures: ['local'] } )
  .then(onInlineSessionCreated);
}

function onVRClick() {
  xr.requestSession('immersive-vr')
  .then(onVRSessionCreated);
}
```

In response, the UA must ensure that:
* The document is allowed to use all the policy-controlled features associated with the sensor types used to track the native origin of a `local` reference space

> * If the session mode is `inline`, user intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

Reviewer 3: typo in link
```suggestion
* If the session mode is `inline`, user intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implied-consent)
```

}
```

In addition to meeting the [`XRPose`](#xrpose) requirements, every call to `XRFrame.getViewerPose()` which will return more than one `XRView` must additionally ensure that:
* User intention is well understood, either via [explicit consent](#explicit-consent) or [implicit consent](#implict-consent)

> * If `XRView` data is affected by settings that may vary from device to device, such as static interpupillary distance or variations in screen geometry, then the XRView data must be anonymized to prevent fingerprinting. Specific approaches to this are at the discretion of the user agent.

Reviewer 3: Anonymization only prevents fingerprinting (not user profiling). The profiling threat is addressed through user consent. To clarify the actual profiling threat, we should add text and I've added #748 for that.

In the meantime, this line and the following could probably be combined. Suggested change for this line (and the following one removed):

```suggestion
* If `XRView` data is affected by settings that may vary from device to device, such as static interpupillary distance, variations in screen geometry, or user-configured interpupillary distance, then the XRView data must be anonymized to prevent fingerprinting. Specific approaches to this are at the discretion of the user agent.
```

Author: I'm a bit confused.  I separated these lines because I thought we needed to categorize the profiling threat separately from the fingerprinting threat.  Is that not the case?

Reviewer 3: We do, but anonymization (e.g. rounding) only addresses the fingerprinting risk. We can collapse all the fingerprinting mitigations into  a single line since they're the same.

For profiling risk, the mitigation is informed user consent. I've got suggested text in a comment for #748

Author: Ah, ok, so this is a bit tricky, because the informed user consent would need to be requested at the time the reference space is created not at the time pose data is queried.  (That's actually one of the reasons I didn't just go ahead and merge your proposed changes in #748 into this PR).  I can collapse these two sentences for now, but we'll need to make sure any additional text regarding consent is added to the correct place in the document.  Sound good?

Author: Actually... what about this:
>* If `XRView` data is affected by a user-configured interpupillary distance, then it is strongly recommended that explicit consent be requested for the creation of the `XRReferenceSpace` passed into `XRFrame.getViewerPose()`.

Reviewer 3: SGTM. Can you outline your concerns about what could be tricky in #748?

> * If `XRView` data is affected by a user-configured interpupillary distance, then the XRView data must be anonymized to prevent profiling as well as fingerprinting. Specific approaches to this are at the discretion of the user agent.

Reviewer 3: With suggestion on previous line, I believe this one can be removed as anonymization prevents fingerprinting but not profiling.

Author: I believe this is fixed now. Let me know if not!
